Dual stack 1.20 docs update

bridgetkromhout opened the pull request:
Dual stack services PR: https://github.com/kubernetes/kubernetes/pull/91824
Dual stack support enhancement issue: https://github.com/kubernetes/enhancements/issues/563

Please review.

Signed-off-by: Bridget Kromhout <bridget@kromhout.org>
Co-authored-by: Tim Bannister <tim@scalefactory.com>
Co-authored-by: Lachlan Evenson <lachlan.evenson@microsoft.com>

bridgetkromhout commented:
/milestone 1.20

annajung commented:
/milestone 1.20
/sig network
/hold 
pending merge of kubernetes/kubernetes#91824
/cc @reylejano-rxm

sftim commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -15,7 +15,7 @@ weight: 70
 <!-- overview -->
 
-{{< feature-state for_k8s_version="v1.16" state="alpha" >}}
+{{< feature-state for_k8s_version="v1.20" state="alpha" >}}
```
BTW it's actually OK to leave this line unchanged.

sftim replied:
In the text of the page, signpost readers to read the version of this page for _their_ version of Kubernetes - so that they are viewing accurate documentation for this alpha feature.

lachie83 replied:
This is an unusual case as this enhancement has been reimplemented in Kubernetes 1.20 and any docs prior to 1.20 (1.16 though 1.19) are incorrect. Given that, how do you suggest we signal that these docs only are relevant to this version of this alpha enhancement and not the prior implementation?

sftim replied:
My suggestion for this is / was: https://github.com/kubernetes/website/pull/24725#discussion_r516152624

If people are using v1.18 and read the v1.18 docs, the two should match.

annajung commented:
/hold cancel

khenidak commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -30,14 +31,14 @@ If you enable IPv4/IPv6 dual-stack networking for your Kubernetes cluster, the c
 The following prerequisites are needed in order to utilize IPv4/IPv6 dual-stack Kubernetes clusters:
 
-   * Kubernetes 1.16 or later
+   * Kubernetes 1.20 or later (Dual-stack as alpha was re-implemented in 1.20)
```
This might confuse some people, we only re implemented the service part.

bridgetkromhout replied:
Okay. How do you suggest we communicate that these docs only apply in full to the 1.20 version, and may be incorrect for the first alpha? (As far as I understand it, we do not intend to maintain docs for the first alpha implementation.)

sftim replied:
My suggestion for this is / was: https://github.com/kubernetes/website/pull/24725#discussion_r516152624

bridgetkromhout replied:
We've implemented @sftim's suggestion.

khenidak commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -68,43 +69,60 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+Services will default to the address family of the first service cluster IP range configured  via the `--service-cluster-ip-range` flag to the kube-controller-manager.
 
-You can set `.spec.ipFamily` to either:
+If you want to explicitly specify which IP address is the primary IP address (which is the address assigned for the Service as `.spec.ClusterIP`), you can optionally use the `.spec.ipFamilyPolicy` field. This field accepts one of the following values:
```
`s/IP address/IP address family/`

Not to be confused with user ability to select an actual IP for their services.

bridgetkromhout replied:
We now say:
```
If your cluster has dual-stack enabled, you can create {{< glossary_tooltip text="Services" term_id="service" >}} which can use IPv4, IPv6, or both. 

The address family of a Service defaults to the address family of the first service cluster IP range (configured via the `--service-cluster-ip-range` flag to the kube-controller-manager).
```

khenidak commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -68,43 +69,60 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
-{{< note >}}
-The default address family for your cluster is the address family of the first service cluster IP range configured via the `--service-cluster-ip-range` flag to the kube-controller-manager.
-{{< /note >}}
+Services will default to the address family of the first service cluster IP range configured  via the `--service-cluster-ip-range` flag to the kube-controller-manager.
```
This does not cover:

1. Existing Services (default to single stack + family of whatever IP assigned to that service)
2. Existing services (headless w/selector)
3. The ability to upgrade and downgrade services.

lachie83 replied:
Thanks Kal. We will work on added these sections

bridgetkromhout replied:
Added as requested

sftim commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -68,43 +69,60 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+{{< codenew file="service/networking/dual-stack-default-svc.yaml" >}}
 
-### Type LoadBalancer
+   * This Service specification explicitly defines `PreferDualStack` in `.spec.ipFamilyPolicy`. Kubernetes will assign both IPv4 and IPv6 addresses (as this cluster has dual-stack enabled) and select the `.spec.ClusterIP` from the list of `.spec.ClusterIPs` based on the address family of the first element in the `.spec.ipFamilies` array. Specifying `RequireDualStack` in `.spec.ipFamilyPolicy` on  ...
```
Is this accurate?
```suggestion
This Service specification explicitly defines `PreferDualStack` in `.spec.ipFamilyPolicy`.
When you create this Service on a dual-stack cluster, Kubernetes assigns both IPv4 and IPv6 addresses for the service. The control plane updates the `.spec` for the Service to record the IP address assignments. The field `.spec.ClusterIPs` is the primary record, and contains both assigned IP addresses; `.spec.ClusterIP` is a secondary field with its value calculated from `.spec.ClusterIPs`.  
For the `.spec.ClusterIP` field, the control plane records the IP address that is the same address family as the cluster's main IP address family. 

On a single-stack cluster, the `.spec.ClusterIPs` and `.spec.ClusterIP` fields both only list one address.
On a cluster with dual-stack enabled, specifying `RequireDualStack` in `.spec.ipFamilyPolicy` behaves the same as `PreferDualStack`.
```

sftim commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -68,43 +69,60 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+
+## Egress Traffic
 
+Either publicly routable or non-publicly routable IPv6 address blocks will work if the underlying {{< glossary_tooltip text="CNI" term_id="cni" >}} provider implements the transport. If you have a Pod that uses non-publicly routable IPv6 addresses and want that Pod to reach off-cluster destinations (eg. the public Internet), you must set up IP masquerading for the egress traffic. The [ip-masq-age ...
```
Is masquerading the only option?
```suggestion
Either publicly routable or non-publicly routable IPv6 address blocks will work if the underlying {{< glossary_tooltip text="CNI" term_id="cni" >}} provider implements the transport. If you have a Pod that uses non-publicly routable IPv6 addresses and want that Pod to reach off-cluster destinations (eg. the public Internet), you must set up IP masquerading for the egress traffic, or a similar mechanism such as transparent proxying. The [ip-masq-agent](https://github.com/kubernetes-sigs/ip-masq-agent) project supports dual-stack for IP masquerading on dual-stack clusters.
```

sftim commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -68,43 +69,60 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+   * Set `.spec.ipFamilyPolicy` field to `PreferDualStack` or `RequireDualStack`
+
+{{< note >}}
+IPv6-enabled external load balancers must be supported by your cloud provider.
```
```suggestion
To use a dual-stack `LoadBalancer` type Service, your cloud provider must support IPv4 and IPv6 load balancers.
```
?

sftim commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -68,43 +69,60 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+### Headless Services without selector
 
-## Known Issues
+For [Headless Services without selectors](/docs/concepts/services-networking/service/#without-selectors) and without `.spec.ipFamilyPolicy` explicitly set, the `.spec.ipFamilyPolicy` field will default to `RequireDualStack`.
```
Wording nit
```suggestion
For [Headless Services without selectors](/docs/concepts/services-networking/service/#without-selectors) and without `.spec.ipFamilyPolicy` explicitly set, the `.spec.ipFamilyPolicy` field defaults to `RequireDualStack`.
```

sftim reviewed (commented):
Hi @bridgetkromhout 

I hope this feedback is helpful. Some details are just style guide suggestions and are optional. Other suggestions I made are aiming to help the reader understand dual-stack operation.

sftim commented:
/hold
Please squash commits when ready (BTW this repo preserves “LGTM” labels through squashes) and feel free to remove this hold once squashed.

irvifa commented:
/label tide/merge-method-squash

lachie83 commented:
/remove-language id

annajung commented:
/assign @reylejano-rxm

bridgetkromhout commented:
/remove-language id

lachie83 commented:
/remove-language id
/remove-language ko
/remove-language ja
/remove-language fr
/remove-language zh

bridgetkromhout commented:
/unhold

sftim commented:
> This is a placeholder PR to be updated.

@bridgetkromhout please edit the PR description to clarify. If it is still WIP, please retitle or otherwise relabel it as work-in-progress.

bridgetkromhout commented:
/remove-language id

bridgetkromhout commented:
/remove-language id

annajung commented:
@kubernetes/sig-network-pr-reviews Can you provide a technical review (lgtm)?

bridgetkromhout commented:
Hi, @sftim - In this docs update, we removed a file that's no longer valid.

* Unfortunately, the build process doesn't work without the file (even though, in theory, nothing else should be referencing the `en` version of it. So, I replaced the file left those commits unsquashed - maybe you can advise us on how to remove the file?

* Weirdly, the PR keeps being marked by the bot as being `language/id`. It definitely is not Indonesian, but the lack of that file in the `language/id` directory also breaks the build.

Please help sort this out so we can get this PR finished - thanks!

khenidak commented:
/lgtm

raghvenders commented:
@annajung   -  As you mentioned about the issue in slack , i got a chance to run locally as changing the indonesian (id) file - id/docs/concepts/services-networking/dual-stack.md - "dual-stack-ipv6-svc.yaml" which is removed though and replacing it with "dual-stack-preferred-svc.yaml" makes the build success.

But @bridgetkromhout is not looking for changes in Indonesian based on above comments but contrary to that i could see some of her commits for yaml file under /id folder though.

> Hi, @sftim - In this docs update, we removed a file that's no longer valid.
> 
> * Unfortunately, the build process doesn't work without the file (even though, in theory, nothing else should be referencing the `en` version of it). So, I replaced the file and left those commits unsquashed - maybe you can advise us on how to remove the file?

I could see more issue under - https://raw.githubusercontent.com/kubernetes/website/master/content/en/examples/service/networking/dual-stack-preferred-ipfamilies-svc.yaml give 404 when you click on the link in "dual-stack-preferred-ipfamilies-svc.yaml".I believe this was the issue @bridgetkromhout was referring above?

> * Weirdly, the PR keeps being marked by the bot as being `language/id`. It definitely is not Indonesian, but the lack of that file in the `language/id` directory also breaks the build.

I am unable to to see the label as `language/id`
> 
> Please help sort this out so we can get this PR finished - thanks!

bridgetkromhout commented:
Hi, @raghvenders -

> based on her commits ,I could see some of her commits for yaml file under /id folder though which needs respective reference.

You're right! I am not sure how that happened. I'll correct that tomorrow. =)

raghvenders commented:
@bridgetkromhout  - for 404 on the updated link.Hope it is testable once it got deployed not from local because codenew looks like served from "# See codenew shortcode
githubWebsiteRaw = "raw.githubusercontent.com/kubernetes/website""

@sftim  - please do correct me here.

bridgetkromhout commented:
/remove-language id

kbhawkey commented:
Since there is one commit,
/remove-label tide/merge-method-squash

kbhawkey commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -68,47 +72,162 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+When you define a Service you can optionally configure it as dual stack. To specify the behavior you want, you
+set the `.spec.ipFamilyPolicy` field to one of the following values:
+
+   * `SingleStack`: Single-stack service. The control plane allocates a cluster IP for the Service, using the first configured service cluster IP range.
```
nit: I think you do not need to indent the list.

kbhawkey commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -68,47 +72,162 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+
+#### Dual-stack options on new Services
+
+   1. This Service specification does not explicitly define `.spec.ipFamilyPolicy`. When you create this Service, Kubernetes assigns a cluster IP for the Service from the first configured `service-cluster-ip-range` and sets the `.spec.ipFamilyPolicy` to `SingleStack`. ([Services without selectors](/docs/concepts/services-networking/service/#services-without-selectors) and [headless Services](/doc ...
```
You can remove the indentation of this list.
If you want a numbered list, you can use:

1. item 1
1. item 2
1. item 3

sftim replied:
The source for that Markdown would look like
```markdown
1. item 1
1. item 2
1. item 3
```

bridgetkromhout replied:
Enough elements between that 1. 1. 1. doesn't work in this case.

kbhawkey commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -68,47 +72,162 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+{{< codenew file="service/networking/dual-stack-preferred-svc.yaml" >}}
 
-For comparison, the following Service specification will be assigned an IPv4 address (also known as a "cluster IP") from the configured `service-cluster-ip-range` to this Service.
+   3. This Service specification explicitly defines `IPv6` and `IPv4` in `.spec.ipFamilies` as well as defining `PreferDualStack` in `.spec.ipFamilyPolicy`. Kubernetes will assign an IPv6 and IPv4 address in `.spec.ClusterIPs` and `.spec.ClusterIP` will be set to the IPv6 address because it is the first element in the `.spec.ClusterIPs` array, overriding the default. 
```
What do you think about this edit:
When Kubernetes assigns an IPv6 and IPv4 address in `.spec.ClusterIPs`, the `.spec.ClusterIP` field is set to the IPv6 address because it is the first element in the `.spec.ClusterIPs` array, overriding the default.

kbhawkey commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -68,47 +72,162 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+These examples demonstrate the default behavior when dual-stack is newly enabled on a cluster where Services already exist.
 
-## Egress Traffic
+   1. When dual-stack is enabled on a cluster, existing Services (whether `IPv4` or `IPv6`) are configured by the control plane to set `.spec.ipFamilyPolicy` to `SingleStack` and set `.spec.ipFamilies` to the address family of the existing Service. The existing Service cluster IP will be stored in `.spec.ClusterIPs`. 
```
You can remove the indentation?

kbhawkey commented on content/en/docs/concepts/services-networking/dual-stack.md:
````diff
@@ -68,47 +72,162 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+```yaml
+kubectl get svc my-service -o yaml
+apiVersion: v1
+kind: Service
````
nit: Is this YAML the output from the `kubectl get` command?

bridgetkromhout replied:
We revised a bit, but this does need to be explained.

kbhawkey commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -68,47 +72,162 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+
+Services can be changed from single-stack to dual-stack and from dual-stack to single-stack.
+
+   1. To change a Service from single-stack to dual-stack, change `.spec.ipFamilyPolicy` from `SingleStack` to `PreferDualStack` or `RequireDualStack` as desired. When you change this Service from single-stack to dual-stack, Kubernetes assigns the missing address family so that the Service now has IPv4 and IPv6 addresses.
```
nit: remove indentation of list

kbhawkey commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -68,47 +72,162 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+
+For [Headless Services without selectors](/docs/concepts/services-networking/service/#without-selectors) and without `.spec.ipFamilyPolicy` explicitly set, the `.spec.ipFamilyPolicy` field defaults to `RequireDualStack`.
+
+### Service Type LoadBalancer
```
nit: Service type load balancer?

bridgetkromhout replied:
As we're talking about a Service of this LoadBalancer type, we tried to keep the capitalization as it is in the other docs and code: https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer

kbhawkey commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -68,47 +72,162 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+
+### Service Type LoadBalancer
+
+To provision a dual-stack LoadBalancer for your Service:
```
nit: load balancer vs LoadBalancer?

bridgetkromhout replied:
See above - https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer

kbhawkey commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -68,47 +72,162 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+{{< /note >}}
 
-## {{% heading "whatsnext" %}}
+## Egress Traffic
```
s/Egress Traffic/Egress traffic

kbhawkey commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -68,47 +72,162 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
-## {{% heading "whatsnext" %}}
+## Egress Traffic
 
+Either publicly routable or non-publicly routable IPv6 address blocks will work if the underlying {{< glossary_tooltip text="CNI" term_id="cni" >}} provider implements the transport. If you have a Pod that uses non-publicly routable IPv6 addresses and want that Pod to reach off-cluster destinations (eg. the public Internet), you must set up IP masquerading for the egress traffic, or a similar mec ...
```
nit: What are the other options that "work" besides publicly or non-publicly routable?
In other words, could you say, "all routable IPv6 address blocks are routed if the underlying ..."

bridgetkromhout replied:
Okay, have clarified and condensed this point. Thanks.

kbhawkey commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -68,47 +72,162 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+  name: my-service
+spec:
+  clusterIP: None
+  clusterIPs:
```
nit: Do you need to show the full YAML?

bridgetkromhout replied:
We think it's needed here for clarity.

kbhawkey commented on content/en/docs/tasks/network/validate-dual-stack.md:
````diff
@@ -134,28 +137,89 @@ status:
+The `kubectl get svc` command will only show the primary IP in the `CLUSTER-IP` field.
+
+```shell
+$  kubectl get svc -l app=MyApp
````
nit: You can remove the "$" prompt.

sftim replied:
Better to separate the command from its output (two blocks)

kbhawkey commented on content/en/docs/tasks/network/validate-dual-stack.md:
```diff
@@ -134,28 +137,89 @@ status:
+{{< codenew file="service/networking/dual-stack-ipfamilies-ipv6.yaml" >}}
 
-Validate that the Service gets a cluster IP address from the IPv6 address block. You may then validate access to the service via the IP and port.
+By viewing the YAML for the Service, observe that the Service has the `.spec.ipFamilyPolicy` field has set to `SingleStack` and the `.spec.clusterIP` set to an IPv6 address from the IPv6 range set via `--service-cluster-ip-range` flag on kube-controller-manager.
```
Read through line 144 and reword.

kbhawkey commented:
Page previews:
https://deploy-preview-24725--kubernetes-io-vnext-staging.netlify.app/docs/concepts/services-networking/dual-stack/
https://deploy-preview-24725--kubernetes-io-vnext-staging.netlify.app/docs/tasks/network/validate-dual-stack/

sftim commented:
/remove-area blog

sftim commented on content/en/docs/concepts/services-networking/dual-stack.md:
````diff
@@ -68,47 +72,162 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+   You can validate this behavior by using kubectl to inspect an existing service.
+
+```yaml
+kubectl get svc my-service -o yaml
````
BTW this line isn't YAML. See https://kubernetes.io/docs/contribute/style/style-guide/#separate-commands-from-output for how the style guide recommends showing commands and their output.

sftim commented on content/en/docs/concepts/services-networking/dual-stack.md:
````diff
@@ -68,47 +72,162 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+   You can validate this behavior by using kubectl to inspect an existing headless service with selectors.
+
+```yaml
+kubectl get svc my-service -o yaml
````
This line isn't YAML

sftim commented on content/en/docs/tasks/network/validate-dual-stack.md:
```diff
@@ -2,6 +2,7 @@
 - lachie83
 - khenidak
+- bridgetkromhout
 min-kubernetes-server-version: v1.16
```
```suggestion
min-kubernetes-server-version: v1.20
```
?

bridgetkromhout replied:
Yes - done.

sftim commented on content/en/docs/tasks/network/validate-dual-stack.md:
````diff
@@ -97,31 +98,33 @@ a00:100::4    pod01
 ```
 
 ```yaml
+kubectl get svc my-service -o yaml
````
This line isn't YAML.

sftim commented on content/en/docs/tasks/network/validate-dual-stack.md:
````diff
@@ -134,28 +137,89 @@ status:
-my-service   ClusterIP   fe80:20d::d06b   <none>        80/TCP    9s
+
+```yaml
+kubectl get svc my-service -o yaml
````
This line isn't YAML.

sftim commented on content/en/docs/tasks/network/validate-dual-stack.md:
````diff
@@ -134,28 +137,89 @@ status:
+
+```shell
+kubectl describe svc -l app=MyApp
+Name:              my-service
+Namespace:         default
+Labels:            app=MyApp
+Annotations:       <none>
+Selector:          app=MyApp
+Type:              ClusterIP
+IP Family Policy:  PreferDualStack
+IP Families:       IPv4,IPv6
+IP:                10.0.216.242
+IPs:               10.0.216.242,fd00::af55
+Port:              <unset>  80/TCP
+TargetPort:        9376/TCP
+Endpoints:         <none>
+Session Affinity:  None
+Events:            <none>
````
This text is not valid Bourne-style shell.

sftim reviewed (commented):
Documentation LGTM. Some style nits, I can pick those up once this merges.
Ready for tech review from SIG Network.

reylejano-rxm commented:
@kubernetes/sig-network-pr-reviews can you provide a tech review?

kbhawkey commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -68,47 +72,169 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
 {{< codenew file="service/networking/dual-stack-default-svc.yaml" >}}
 
-The following Service specification includes the `ipFamily` field. Kubernetes will assign an IPv6 address (also known as a "cluster IP") from the configured `service-cluster-ip-range` to this Service.
+1. This Service specification explicitly defines `PreferDualStack` in `.spec.ipFamilyPolicy`. When you create this Service on a dual-stack cluster, Kubernetes assigns both IPv4 and IPv6 addresses for the service. The control plane updates the `.spec` for the Service to record the IP address assignments. The field `.spec.ClusterIPs` is the primary field, and contains both assigned IP addresses; `. ...
```
Hi @bridgetkromhout .
If you have time, I think the list numbering will work if you indent the "codenew" shortcode (4 spaces).

bridgetkromhout replied:
Thanks. Let's catch the refactoring after we get all the content locked!

raghvenders replied:
@kbhawkey @bridgetkromhout - i saw some lints mentioned in the project structure.will one such exists help on these here?

khenidak commented:
/lgtm

thockin reviewed (commented):
Thanks!

/lgtm
/approve

annajung commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -68,47 +72,173 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+{{< codenew file="service/networking/dual-stack-preferred-svc.yaml" >}}
 
-For comparison, the following Service specification will be assigned an IPv4 address (also known as a "cluster IP") from the configured `service-cluster-ip-range` to this Service.
+1. This Service specification explicitly defines `IPv6` and `IPv4` in `.spec.ipFamilies` as well as defining `PreferDualStack` in `.spec.ipFamilyPolicy`. When Kubernetes assigns an IPv6 and IPv4 address in `.spec.ClusterIPs`, `.spec.ClusterIP` is set to the IPv6 address because that is the first element in the `.spec.ClusterIPs` array, overriding the default.
```
Numbers are not incrementing as expected as shown here https://5fab100b181c9c0007af911f--kubernetes-io-vnext-staging.netlify.app/docs/concepts/services-networking/dual-stack/

Could you change to make sure page generates as expected?

bridgetkromhout replied:
Thanks. I think that's one of the stylistic changes @sftim intends to reformat.

bridgetkromhout commented:
/assign sftim

irvifa reviewed (commented):
/approve

reylejano-rxm commented:
Hi @bridgetkromhout @lachie83 , since this feature will be alpha for 1.20 does this feature need to be added to the [Feature Gates list](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-gates-for-alpha-or-beta-features)

irvifa commented:
@reylejano-rxm should we revert the PR first or make a follow up PR will suffice to fix this?
I’m sorry I forgot about the feature gate list.

kbhawkey commented:
Hi @irvifa .
I am guessing that there will be a follow-up PR. I suggest to leave this PR merged.
There is time to fix a few remaining docs issues before the release.

sftim commented:
https://kubernetes.io/docs/concepts/services-networking/dual-stack/ already mentions feature gates, right?

sftim commented:
https://kubernetes.io/docs/concepts/services-networking/dual-stack/ already mentions feature gates, right?

raghvenders commented:
@sftim  , as in the conversation this is mentioned as alpha v1.20 but i still see 

https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-gates-for-alpha-or-beta-features - 


**IPv6DualStack | false | Alpha | 1.16
-- | -- | -- | --**


it needs to be changed to V1.20 and that alpha-beta features list link to be updated in  "https://kubernetes.io/docs/concepts/services-networking/dual-stack/"

doest that sound correct?

sftim commented:
IPv6DualStack has been in alpha since v1.16 and is still in alpha. https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-gates-for-alpha-or-beta-features looks right to me.
